### PR TITLE
Add tag flavor

### LINKED ISSUE #13: Add appending option to tags
Let's say you build multiple distro flavors of your Docker image, e. g. one in Ubuntu (main) and one in Alpine.
Meanwhile it's fine to use the usual tagging for the main image, is there a way of setting a "base tag" or a "tag" to be appended? In this case, for the Alpine flavor, instead of `latest`, using `alpine`, `x.x.x-alpine` etc. is more appropriate.

### PR DESCRIPTION
**Pull request by hugopeixoto**
Fixes #13 

For cases where you have multiple flavors of the same container, like `node:12-alpine`, `node:12-debian`, etc, it would be nice to be able to specify these flavors in this github action. This was requested in #13.

Let's say that your container has two flavors, `debian` and `alpine`, and that you want images like `foo:latest` to be based on the debian flavor. You might want to generate these tags:

1. `foo:x.y.z-debian`
1. `foo:x.y.z-alpine`
1. `foo:x.y.z` (`= foo.x.y.z-debian`)
1. `foo:debian` (`= foo.x.y.z-debian`)
1. `foo:alpine` (`= foo.x.y.z-alpine`)
1. `foo:latest` (`= foo.x.y.z-debian`)
1. `foo:sha-AAAAAA-debian`
1. `foo:sha-AAAAAA-alpine`
1. `foo:sha-AAAAAA` (`= `foo:sha-AAAAAA-debian`)

To achieve this, I added two options: `flavor (string)` and `main-flavor (boolean)`. Trying to summarize their behavior:
- specifying a flavor will generate tags with the flavor suffix
- if no flavor is specified, or if there is a flavor but `main-flavor` is true, the flavorless tags will be emitted as well
- versionless flavorful tags (`foo:debian`) will only be generated if there's a flavor and the tag matches as `latest`.

In other words, each previously emited tag has its flavor counterpart:
- `foo:x.y.z` => `foo:x.y.z-{flavor}`
- `foo:latest` => `foo:{flavor}`
- `foo:sha-AAAAAA` => `foo:sha-AAAAAA-{flavor}`

And the left side are emited `if no flavor != '' or mainFlavor`, and the right side is emited `if flavor != ''`.

This is how I'm using this in a project:

```yaml
on:
  push:
    tags: '*.*.*'

jobs:
  release:
    strategy:
      matrix:
        base: [debian, alpine]
    steps:
      - uses:  hugopeixoto/ghaction-docker-meta@d08260d
        name: Docker meta
        id: docker_meta
        with:
          images: quay.io/hedgedoc/hedgedoc
          flavor: ${{ matrix.base }}
          main-flavor: ${{ matrix.base == 'debian' }}

```

I hope this is understandable :P Let me know what you think, and if this is an approach that makes sense. I wasn't sure if the `sha` tags should be flavored as well.

I still need to add tests and document this in the readme, but I'd like to know if this is something that you would consider integrate and iterate on before spending time there.

### CODE REVIEW

**Comment by crazy-max** [comment]
@hugopeixoto Thanks for this PR. Can you merge please?

**Comment by hugopeixoto** [comment]
> @hugopeixoto Thanks for this PR. Can you merge please?

I've rebased the PR to the latest master, resolving the conflicts. Let me know if you meant something else.

**Comment by crazy-max** [comment]
@hugopeixoto Ahah my bad was tired yesterday, I meant rebase :sweat_smile:

**Comment by LongLiveCHIEF** [comment]
@hugopeixoto you beat me to it.

File: src/meta.ts [hunk 1]
```diff
@@ -94,18 +94,26 @@ export class Meta {
...
       }
       if (version.latest) {
-        tags.push(`${imageLc}:latest`);
+        if (main) tags.push(`${imageLc}:latest`);
+        if (hasFlavor) tags.push(`${imageLc}:${flavor}`);
       }
       if (this.context.sha && this.inputs.tagSha) {
-        tags.push(`${imageLc}:sha-${this.context.sha.substr(0, 7)}`);
+        if (main) tags.push(`${imageLc}:sha-${this.context.sha.substr(0, 7)}`);
+        if (hasFlavor) tags.push(`${imageLc}:sha-${this.context.sha.substr(0, 7)}-${flavor}`);
```
Comment (crazy-max) [inline src/meta.ts:116]:
```suggestion
        if (main) {
          tags.push(`${imageLc}:sha-${this.context.sha.substr(0, 7)}`);
        }
        if (flavor) {
          tags.push(`${imageLc}:sha-${this.context.sha.substr(0, 7)}-${flavor}`);
        }
```

**Review by crazy-max — CHANGES_REQUESTED** [review]
Overall LGTM but need some changes:

* Please don't create inline condition. We want to make sure code coverage is respected.
* Add tests
* Update README

**Comment by yumetodo** [comment]
I want to append suffix(like `-tesseract_opencv-4.5.0`). When this PR is merged, can I do that?

**Comment by LongLiveCHIEF** [comment]
I suspect #24 may supersede this entirely, even though it may not have been intentionally designed to do so.

**Comment by crazy-max** [comment]
@LongLiveCHIEF 

> I suspect #24 may supersede this entirely, even though it may not have been intentionally designed to do so.

Actually no, `tag-custom` add additional tags to those already produced and do not modify them as this PR does.

**Comment by hugopeixoto** [comment]
I've updated the PR to include documentation in README.md and some tests in meta.test.ts.

@crazy-max I'm fighting codecov and losing. The current problem seems to be the line `let main = !flavor || this.inputs.mainFlavor;`, which is marked as partial, but I think I have the four cases covered. Could you help me figure out what's missing? The coverage report from `yarn test` doesn't seem to be as extensive as whatever codecov does.

**Comment by yumetodo** [comment]
How about this? Some coverage tools cannot recognize short circuits correctly.

```ts
    let tags: Array<string> = [];
    const tagPush = (tagStr: string) => {
      if (this.inputs.flavor) {
        tags.push(`${tagStr}-${this.inputs.flavor}`);
        if (this.inputs.mainFlavor) {
          tags.push(tagStr);
        }
      } else {
        tags.push(tagStr);
      }
    };
    for (const image of this.inputs.images) {
      const imageLc = image.toLowerCase();
      tagPush(`${imageLc}:${this.version.main}`);
      for (const partial of this.version.partial) {
        tagPush(`${imageLc}:${partial}`);
      }
      if (this.version.latest) {
        if (this.inputs.flavor) {
          tags.push(`${imageLc}:${this.inputs.flavor}`);
          if (this.inputs.mainFlavor) {
            tags.push(`${imageLc}:latest`);
          }
        } else {
          tags.push(`${imageLc}:latest`);
        }
      }
      if (this.context.sha && this.inputs.tagSha) {
        tagPush(`${imageLc}:sha-${this.context.sha.substr(0, 7)}`);
      }
    }
```

**Comment by hugopeixoto** [comment]
> How about this? Some coverage tools cannot recognize short circuits correctly.

Thanks for the suggestion. I tried a variation of that: instead of using `let main = x || y`, I tried this:

```typescript
if (flavor === '') {
  main = true;
} else if (this.inputs.mainFlavor) {
  main = true;
} else {
  main = false;
}
```

And each of the `main = true/false` lines is green (hit), but the if/else lines are yellow (partial). It seems like every `if` statement in this file is being marked as partial:
https://codecov.io/gh/crazy-max/ghaction-docker-meta/pull/15/src/src/meta.ts

Could it be that there's a problem with codecov's typescript tooling? I rolled the commit back to `!flavor || etc`.

**Comment by yumetodo** [comment]
Oh...

**Comment by crazy-max** [comment]
@hugopeixoto Yes don't worry with codecov for this kind of stuff, there's something wrong on their side about it or something needs to be configured on our side. Will check this out later.

**Comment by yumetodo** [comment]
@crazy-max Do you have any updates? Are there any other problems?

**Comment by felbinger** [comment]
Exactly the feature I need :+1:

**Comment by lpil** [comment]
Hi @crazy-max , is there more to do here? 🙏

**Comment by crazy-max** [comment]
@lpil Will focus on proposal in #30 and also work to drastically reduce the workflow by removing a maximum of inputs. I keep this PR opens for now as it provides a good use case.

**Pull request closed without merging**